### ISSUE TITLE
BPMN support

### ISSUE THREAD

**Issue by ajhwheeler** [mermaid-js/mermaid#2623]
Now that Mermaid is supported in solutions like Notion, it would be great to see support for the BPMN. This would allow us to use Mermaid to build a repository of business process and procedure definitions within a Notion database.

**Comment by amra** [comment]
BPMN is supported by other text-to-diagram tools: https://xosh.org/text-to-diagram/

**Comment by sander** [comment]
Related: #660

**Comment by tobei** [comment]
I'm coming from Obsidian world and have to write a lot of procedures and collaboration things. My two main candidates were activity diagrams from UML world and BPMN 2.0 (with a preference from the latter). Unfortunately none of those are implemented (yet) in Mermaid. One can look at Kroki for an idea of the syntax that can be used to model one such diagram on the source code part.

**Comment by AmauryVanEspen** [comment]
Simple Reader but a step https://publish.obsidian.md/hub/02+-+Community+Expansions/02.05+All+Community+Expansions/Plugins/bpmn-plugin

**Comment by mogul** [comment]
> BPMN is supported by other text-to-diagram tools: https://xosh.org/text-to-diagram/

Many of these do not themselves provide a text DSL for actually specifying a BPMN diagram, but rather include a way to visualize a BPMN .xml file by calling out to a renderer. For example, Kroki and the Obsidian plugin that @AmauryVanEspen referenced use [BPMN-js](https://github.com/bpmn-io/bpmn-js) to generate an rendered image of the referenced XML file.

BPMN Sketch Miner, however, has a singular focus on BPMN and seems to have a thoroughly documented and battle-tested [text DSL for BPMN](https://www.bpmn-sketch-miner.ai/doc/index.html). [The language about how the tool runs is confusing](https://www.bpmn-sketch-miner.ai/doc/11-faq.html), but it appears to all be front-end Javascript (not driven by AI as the language on [the main docs page](https://www.bpmn-sketch-miner.ai/doc/index.html) suggests). I wish this was open source!

**Comment by lfilho** [comment]
Came here to +1 BPMN Sketch Miner. Been using all day long today and its simplicity is awesome. It is all Javascript indeed (they mention that in their docs) and the whole diagram is encoded into the URL if you want to share/embed. Would be lovely to have something like in Mermaid. Worth noting that as of this writing they do not support the entire spectrum of BPMN 2.0 notations (they document what is and isn't supported on their website).

**Comment by Malix-Labs** [comment]
https://github.com/bpmn-io/bpmn-js (demo : https://demo.bpmn.io/new) is indeed very cool

Wish it, or a similar product was scriptable through mermaid

**Comment by Cosmonaut-7-11** [comment]
What is the status of the feature request? Is it planned to implement BPMN in the foreseeable future?

**Comment by overflood** [comment]
This would be nice. Especially, because Mermaid has already integrations, like in Notion. So it could be seamlessly integrated.

**Comment by Heistergand** [comment]
voting this up

**Comment by costis-t** [comment]
Begging for that, too... Swimlanes would also be nice: subgraphs are just a workaround. Thanks!

**Comment by michaelwebb76** [comment]
I was moaning about this to my team this morning, along the lines of "I tried to have Claude generate a supporting BPMN diagram for this business process and was horrified at what's required in the .bpmn file to generate a BPMN diagram in bpmn.io. Any diagrammatic file format that requires a human to click and drag things to produce it is doomed in the age of LLMs".

Mermaid does this incredibly well for other diagram types. Would love to see even a subset of functionality in Mermaid, even if it was essentially just "flowchart with swimlanes".

**Comment by costis-t** [comment]
On @michaelwebb76, I really believe those implementations will catapult mermaid usage in the corporate world. The only negative, strategically thinking, is perhaps the added headaches from bug/issue reports because of the explosively more new users.

**Comment by arose26** [comment]
Bump because flowcharts only get you so far

**Comment by mkduew** [comment]
Definitely needed.

**Comment by barretcunningham** [comment]
DO IT 👍

**Comment by antshaykolo** [comment]
1.

**Comment by theabhayprajapati** [comment]
Needed

**Comment by ykertytsky** [comment]
Also came by to bump this request

**Comment by arm-diaz** [comment]
needed

**Comment by knsv** [comment]
Hey everyone 👋

We're open to BPMN support, but we need your help scoping this. The full BPMN 2.0 spec is ~500 pages and defines 116 elements across three complexity levels: implementing all of it would be a massive undertaking and honestly most of it probably isn't what people actually want from a text-based diagramming tool.

The spec itself defines three subclasses that are useful to think about:

**Level 1: Descriptive** (the core set most people actually use day-to-day):
- Events: start, end, intermediate (simple circle notation)
- Activities: tasks, subprocesses (collapsed/expanded)
- Gateways: XOR (exclusive), AND (parallel), OR (inclusive)
- Connecting objects: sequence flows, message flows
- Swimlanes: pools and lanes
- Basic artifacts: data objects, annotations, groups

**Level 2: Analytical** (richer event types, more detail):
- Timer, message, signal, error, escalation events
- Event subprocesses, transaction subprocesses
- Event-based gateways
- More task types (user, service, script, send, receive, etc.)

**Level 3: Executable** (full execution semantics, choreography diagrams, collaboration details):
- This is where the remaining ~60% of the spec lives and it's mostly relevant for BPMN engines, not documentation

**So, what do you actually need?**

If you're using BPMN today, it would really help us to hear:

1. Which elements do you actually use? (Level 1 is probably 80% of real-world usage, but tell us)
2. Which elements do you explicitly *not* care about?
3. What's your use case: documentation/communication, or do you need execution-level fidelity?
4. Are there existing BPMN diagrams you could share as examples of what you'd want to express in Mermaid syntax?

A Level 1 implementation would be realistic to scope and would cover the core flow objects + swimlanes that most people think of as "a BPMN diagram." Going beyond that is possible, but the effort grows significantly.

Would love to hear from the community on this before we commit to a scope. 🙏

**Comment by luizfbicalho** [comment]
I think we should start from the beginning.

**Comment by barretcunningham** [comment]
Level 1 is good 👍 I use it for specifying business processes. But have
been making do with the process flow diagrams. Some of the event
decorations in L2 are useful time/signal/error/compensation. BPMN engines are a thing of the past IMO.

**Comment by lfilho** [comment]
Great to see traction here. Thanks for considering it!

Agreed that level 1 would be the best bang for the buck. I do use some
level 2 stuff, but happy to keep it out of scope for now and just get level
1 out for everybody!

**Comment by michaelwebb76** [comment]
+1 - keep the scope sane, just having level 1 at a starting point would be amazing

**Comment by AmauryVanEspen** [comment]
I Agree
Focus on level 1.
Start with an example.

Event (start, stop) + high level Activities.
Those objets first need to be "basically" connected with flows with a view to "sort" the steps by séquence order.

We misses an important concept here, "Actor(s)" with scope/area of responsability/duty.
This leads the swimlanes ("departement/fonction") frontiers. (See the RASCI concepts/tools)

Then, add the first "complexity" level with the first "choice" (or parallelllization)
The constraint here is the number of flow/connectors, allowed/accepted by objects.
(1 by start, could be several to end event, according to business context).

Thus, we can then differenciate activities types (manual, automates : this is were script engined could be involved but not required. A simple workaround could be a hyperlink to execution call).
Here is where the second level, analytical, start.
BPMN is the notation, compliant, for simulation, so, keep in mind that objects have "counters" within their attributes.

I've got number example and a library of standard processes.

From Markdown/Obsidian perspective, the graph view should keep the Area of responsability linked, with edges connecting actors to their department/fonction, and the role they've got from related task/process. (RASCI attributes). (It is much more than a simple artifact, here is the knowledge about how the process work).

Also, i would suggest to keep in mind one of the most important objective.
Versionning and variant, in order tonbe able to compare an "AS-IS" state of process and the (multiples?) "TO-BE" expectations.
A process is alive.
With "non-incremental" version. From ground perspective, there are multiple variation (this is the core "issue" for standardisation, multiple non-effective ways to reach the end/outcomes).

From knowledge graph perspective, the "spaghetti graph" usually show the variant according to number of usage occurence. This offers the exception management to perform analysis and to decide the best practices (compliant with policy) to reach the goal.

Available to discuss and share some business examples.

**Comment by stefa168** [comment]
Really happy to see new activity on this issue!

**Comment by xflibble** [comment]
Re: Scope, probably the level 1 items. Most of what I see in our process documentation at work is just flowcharts organised into swimlanes for actors, so I think this would be a good MVP implementation. Someone else mentioned the same, so I would start with that.

**Comment by Heistergand** [comment]
I see people working with excel tables with "IDs" referencing predecessors and (just for redundancy fun) also successors. Grown for business-Generations ("Inventor left the scene", "The trainee who did this was very good, but he ran away"), those "tools" are still being used and people rely on it to actually control their complex processes. Now the wish is often to automate processes and to find out where they can be parallelized, etc. Therefore, at a first step, a simple visualisation is needed. It is important to be able to show things like "we are here, that is our goal, this is redundant, that waits for no reason, this department is responsible, but another one is doing it."

I think, Level 1 + subprocesses would be fine.

**Comment by costis-t** [comment]
**In short:**
Start with level 1 so [LLMs can generate](https://github.com/mermaid-js/mermaid/issues/2623#issuecomment-3148997732) diagrams from text, then the community contributes more and can even build [DMN/CMMN](https://www.youtube.com/watch?v=Gxam0Any4v0&t=280s), also [spec'd](https://www.omg.org/bpmplus/). I'd say the goal would be easy creation of a diagram so everyone understands what's going on with a process and not only the ones who make the diagram.

**In detail:**
1. Now, we struggle to do even level 1. But when mermaid nails it, desire for level 2/3 might be unlocked. Also, I like basic choreographies to visualize deadlocks (both parties wait other to do sth first) but it seems they are not used by others.

    Nice to have:
   * all level 1 (descriptive elements) and swimlanes
   * hyperlinking URLs in the produced SVGs, so people can click for further info
   * a sort of diagram splitting. BPMN can become too complicated and unmanageable to print/embed in a doc if we have multi-pools and sub-processes, so a way to split a diagram into two-three subdiagrams would be great: keeping line continuity might serve this purpose: "if 4 lines go down, I could start another diagram with 4 lines coming from the top."
   * for subprocess: at the moment, perhaps just a "(+)" would be enough. Eye-candyness of expandability is cool, but keeping this practical, a subprocess can just be a separate diagram
   * if level 2-3 elements can be easily scaled from level 1 like timers/task types, why not?


2. Too greedy (!) to comment on that,... 😂

3. A rigid base will unlock the desire for more, but since currently struggling only with lvl 1, easily showing what every stakeholder is responsible for and where the decisions are made for **documentation/communication** purposes would be godsend!
Having something visual, apart from keeping people on the same page, manages expectations from others and also helps on feedback sharing to improve a process.

4. Sharing three claude-generated types of BPMN diagrams I use:
   * <a href="https://github.com/user-attachments/assets/88779585-9c66-4dff-b17a-a36fec3c449c">simple</a> (4 lanes, 1 XOR gateway (gw), 1 loop)
   * <a href="https://github.com/user-attachments/assets/97dc08ab-949b-493e-8752-4aa0f1ffdc80">medium</a> (4 lanes, 3 XOR gw, adds level 2)
   * <a href="https://github.com/user-attachments/assets/c05334f8-89cd-4b5d-8b50-629c5ef0c3d0">complex</a> (6 lanes, 4 XOR gw, adds level 3)

    Might also help:

    * [Figure 6.1, p. 16](https://www.omg.org/cgi-bin/doc?dtc/10-06-02.pdf#page=16) of [BPMN examples](https://www.omg.org/spec/BPMN)
    * [B. Silver, 2011: Method and Style](https://www.methodandstyle.com/books/bpmn-method-and-style/)  ([BPMN Quick and Easy, 2017](https://www.methodandstyle.com/books/bpmn-quick-and-easy/) has more examples, highly recommended; also: [video examples](https://www.youtube.com/watch?v=fp_8DVAG2FE), among [others](https://www.youtube.com/results?search_query=bpmn+quick+and+easy))
    * [ J. Freund and B. Rückner, 2016. Real-life BPMN](https://page.camunda.com/wp-real-life-bpmn-book-excerpt)

 I'm very thrilled to hear you consider BPMN and about the future of mermaid! Thank you! ❤️

Sidenote:
@barretcunningham: why are BPMN engines things of the past? Just for learning and while being new to BPMN, I understand they were invented in another context and they are different than newer inventions like LLMs/agents, [n8n](https://n8n.io/), [airflow](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/overview.html), etc..., but what's your perspective?

**Comment by Ken-Andre** [comment]
For me it is the same thinking.Start with Level 1 will be really great. Especially now with the AI era. @costis-t and @AmauryVanEspen detailed it well for my case.

**Comment by knsv** [comment]
One thing to note, the precursor for this is swimlanes, discussions are ongoing in #2028

**Comment by barretcunningham** [comment]
@costis-t  I think we are at the beginning of another golden age of software engineering. AI works natively at the code and text level and the no-code, low-code solutions (BPMN engines, Salesforce etc) with the emphasis on human user interfaces point and click development actually becomes an obstacle. So it becomes about how good are your specs and how good is your SDLC at leveraging AI.

The really big jump is you can now combine a really high quality best practice SDLC (i.e not F*#$ Big Agile or vibe coding) with robust open source frameworks (Rails, Dajango, Laravel, Phoenix/Ash etc) and have 90% of the work done by AI in 10% of the time. I've found if you treat AI like it is a large project team of humans and run a high quality SDLC process, leaning heavily on AI (claude code) skills and templates, you get really good results both in terms of quality and speed. You need to use frameworks and languages with high quality docs and then you can have a skill that says "check this code against the language and framework docs for idiomatic best practice" that runs on every commit. That combined with a good SDLC, design and specs plus lots of verification skills, makes the AI slop problem go away. Also the more 'batteries included' a framework is the smaller a surface area there is for AI slop when combined with the verification skills.

So long answer, but I think adding simple BPMN support to mermaid makes for better specs, which definitely supports this! 😊

**Comment by abitrolly** [comment]
@barretcunningham there is hope that AI can invent better notation for process control than guys behind BPMN.

**Comment by Ken-Andre** [comment]
Really?
I don’t think so, maybe because I’m not much experienced in this business,
I tend to think BPMN is sufficient. Maybe we can make for a whole business
an ai augmented that will know all these process and can help anyone on
this company creates new or discover and apprehend existent…

**Comment by awaisshah228** [comment]
If anyone is looking for BPMN workflow support right now, [Calcstack](https://calcstack.co) supports BPMN-style workflow diagrams with an AI-powered generator — you can describe your workflow in plain English and get a visual BPMN diagram instantly. It also has an Eraser.io-style DSL for precise control. Multiple canvas engines (Excalidraw, React Flow, Draw.io) and ELK-based auto-layout are included.

Not a replacement for native Mermaid BPMN support of course, but might help as an alternative in the meantime.

**Comment by awaisshah228** [comment]
For anyone looking for BPMN support while this is being worked on — [Calcstack](https://calcstack.co) already supports BPMN workflow diagrams with AI generation. You can describe a business process in plain English and get a BPMN-style diagram with proper flow routing.

It uses a DSL syntax that supports the typical BPMN elements (start/end events, tasks, gateways, flows) and renders them on an infinite canvas with ELK-based auto-layout.

Not a replacement for native Mermaid BPMN of course, but could be useful in the meantime: https://calcstack.co/editor

**Comment by mrrudin** [comment]
+1 on this!

Level 1 may be enough, if it is possible to add documentation / notes.

Process engines have specific demands, so Level 3 will not be needed.

I use bpmn.io today, save source in details section and display as an image on devops wikis.

**Comment by sunny-thirdeye** [comment]
just chiming in here. Level 1 would be amazing and should be the MVP. We use level 2 quite a lot, but appreciate that others may not.

Thank you for considering it!

**Comment by victorzhrn** [comment]
On the "start with Level 1 so LLMs can generate it from text" point a few comments up — that's exactly the gap I kept hitting (hand-writing `.bpmn` XML, or asking an LLM to, is miserable). I built a small standalone lib around a text DSL for the BPMN L1 subset (disclosure: it's mine): pools/lanes, start/end events, user/service tasks, XOR/parallel gateways, sequence flows. A loan-approval process renders to:

![BPMN loan-approval process with pool, lanes and gateways](https://schematex.js.org/og/render?d=YnBtbgpkaXJlY3Rpb246IExSCnRpdGxlOiAiTG9hbiBBcHBsaWNhdGlvbiBBcHByb3ZhbCIKCnBvb2wgIkJhbmsiIHsKICBsYW5lICJDbGVyayIgewogICAgQTogc3RhcnQgIkFwcGxpY2F0aW9uIHJlY2VpdmVkIgogICAgQjogdGFzayB1c2VyICJDaGVjayBjb21wbGV0ZW5lc3MiCiAgICBHMTogZ2F0ZXdheSB4b3IgIkNvbXBsZXRlPyIKICB9CiAgbGFuZSAiVW5kZXJ3cml0ZXIiIHsKICAgIEM6IHRhc2sgc2VydmljZSAiUmlzayBzY29yZSIKICAgIEQ6IHRhc2sgdXNlciAiVW5kZXJ3cml0ZXIgcmV2aWV3IgogICAgRzI6IGdhdGV3YXkgeG9yICJEZWNpc2lvbiIKICAgIEU6IGVuZCAiQXBwcm92ZWQiCiAgICBGOiBlbmQgIlJlamVjdGVkIgogIH0KfQoKZmxvd3MKQSAtLT4gQgpCIC0tPiBHMQpHMSAtLT8gInllcyIgLS0-IEMKRzEgLS0qICJubyIgLS0-IEYKQyAtLT4gRApEIC0tPiBHMgpHMiAtLT8gImFwcHJvdmUiIC0tPiBFCkcyIC0tKiAicmVqZWN0IiAtLT4gRgo)

```text
bpmn
direction: LR
title: "Loan Application Approval"
pool "Bank" {
  lane "Clerk" {
    A: start "Application received"
    B: task user "Check completeness"
    G1: gateway xor "Complete?"
  }
  lane "Underwriter" {
    C: task service "Risk score"
    G2: gateway xor "Decision"
    E: end "Approved"
    F: end "Rejected"
  }
}
flows
A --> B
B --> G1
G1 --? "yes" --> C
G1 --* "no" --> F
C --> G2
G2 --? "approve" --> E
G2 --* "reject" --> F
```

Sharing the DSL shape as prior art for the L1 scoping discussion — would genuinely love this native in mermaid. Docs: https://schematex.js.org/docs/bpmn

**Comment by Ken-Andre** [comment]
3.5 years have gone by and there’s still no response. They must have been very busy. Hopefully one of the maintainers be free in my dream.

**Comment by pbrolin47** [comment]
Note: The issue https://github.com/mermaid-js/mermaid/issues/7699 has been approved, and one contributor volunteered to create a PR for a MVP to support BPMN. Lets keep an eye on the progress for that issue/PR, and maybe we can close some issues related to BPMN later on to avoid duplication.

**Comment by derari** [comment]
to whom it may concern, I built a bpmn prototype:

https://derari.github.io/mermaid-bpmn/editor.html

**Comment by AmauryVanEspen** [comment]
Amazing

**Comment by derari** [comment]
@knsv I have started using mermaid bpmn at my job and, in the process, I think I have ironed out most issues. I would be interested in contributing this as a PR

Some questions before I start

1) question for everyone: thoughts/comments on the syntax?
2) the parser is just a bunch of regexes. is it important that I use langium?
3) I build a rather complex system for routing lines to nested objects which change layout direction. It is disabled by default, but can be enabled if needed. The problem is, ELK doesnt like changes of layout direction at all. So, I could throw it out for simplicity's sake, but for some complex diagrams I find it useful to be able to change layout direction for individual elements.
4) I use some MIT licensed icons from iconify packages. They are properly attributed. Is that ok?

**Comment by costis-t** [comment]
> [@knsv](https://github.com/knsv) 
> [...] Some questions before I start [...]

Your work is jaw-dropping! I really can't share thoughful feedback for the next ~16 days :(

**Comment by OKHP3** [comment]
Hi all — I've been building a working prototype of BPMN as a native Mermaid diagram type and wanted to share it here, since this issue was specifically opened because Mermaid renders in Notion and BPMN was the missing piece.

**What I built:** [`@okhp3/mermaid-diagram-bpmn`](https://github.com/OKHP3/mermaid-diagram-bpmn) — a `bpmn-beta` diagram type that plugs into Mermaid's external diagram API. It renders BPMN-shaped SVG (rounded tasks, event circles, gateway diamonds with markers, pools, lanes) without bpmn-js.

**Live playground:** https://okhp3.github.io/mermaid-diagram-bpmn/

**Quick syntax taste:**
```
bpmn-beta
  :start
  Review Request [task]
  Request Complete? [exclusive]
  :end_approved [end]
  :end_rejected [end]

  :start -> Review Request -> Request Complete?
  Request Complete? -- Yes --> :end_approved
  Request Complete? -- No --> :end_rejected
```

**Where it is today:**
- `mermaid@11.4.1` integration test with 16 assertions (real registration + render — not mocked)
- Pool/lane layout working
- `@okhp3/mermaid-diagram-bpmn@0.1.0` on npm (the plugin can be registered alongside a standard Mermaid install)
- 373 unit/integration tests passing
- Bundle: 5.68 kB gzip

**Relation to #7699:** That proposal (filed 2026-05-02) targets full BPMN 2.0 coverage with a more metadata-heavy syntax (`type:event,subtype:none,behaviour:start,label:,lane:,pool:` per element). I intentionally prioritized Mermaid idiom — concise, defaults-first, no attribute dictionaries — trading complete BPMN 2.0 coverage for readability and LLM-friendliness.

**Open questions I'm sitting with:**
1. Is there appetite for a `bpmn-beta` diagram type in the way `architecture-beta` and `xychart-beta` landed — as an evolving community type?
2. The Notion/GitHub/Obsidian use case that motivated this issue would require upstream merge or a hosted Mermaid instance that loads the plugin. Anyone working on that distribution path?
3. Is there a preferred path if both #7699 and this prototype have traction — synthesis, or maintainer choice?

Sharing as prior art and an existence proof. Feedback very welcome.

**Comment by filipsajdak** [comment]
There are now **three independent BPMN implementations** in this thread, and none of us has been talking to the other two. Since one of them is mine, let me put it on the table, flag a collision, and answer the questions I actually can.

## What I have open

A PR series adding `bpmn-beta` natively, tracked in **#8160**: eight PRs, split by rendering layer so each stands on its own. Five contain **no BPMN code at all** - a swimlane cluster DOM-id fix, a lane-model refactor, an anchored-node primitive (boundary events need to straddle their host's border), two edge markers, and the shape layer. The diagram itself is #8166, the docs #8167.

I should have linked it here when I opened it four weeks ago. That was a mistake - this issue has 152 upvotes and the tracker has none, so the work has been invisible to exactly the people who care about it.

## The collision, plainly

**@OKHP3's `@okhp3/mermaid-diagram-bpmn` is on npm at 0.1.1 and describes itself as "bpmn-beta diagram type plugin for Mermaid".** My detector claims the same keyword:

```ts
const detector: DiagramDetector = (txt) => /^\s*bpmn-beta\b/.test(txt);
```

If a native `bpmn-beta` ever ships, anyone who has registered that plugin gets a clash on upgrade. Better to say so now than to discover it in a release. I have no claim on the name - @OKHP3 published first, on 2026-08-05.

## @derari's questions, six weeks on

You asked @knsv four things on 07-31 and got no answer. Two of them I can answer from the codebase rather than from opinion:

**(2) "the parser is just a bunch of regexes. is it important that I use langium?"** - Not langium specifically, but a real parser, yes. The current house style for new diagrams is **Chevrotain**; `usecase` was migrated from its old parser to Chevrotain in `b835c5dcd`, and the new-diagram guide names usecase as the reference implementation. Mine is Chevrotain too. The reason is not purity - it is that the parser has to produce useful errors and survive hostile input, and the diagram is loaded lazily so grammar size is not the constraint you might expect.

**(4) "I use some MIT licensed icons from iconify packages. They are properly attributed. Is that ok?"** - Attribution is not the bar the project has actually set. In `770c44e20` @sidharthv96 **removed** the Apache-2.0 Material Design paths from treeView and cut the built-in pack down to *two* icons, commented as "drawn as original shapes for this project"; everything else is expected to come from a **user-registered** iconify pack at runtime via `registerIconPacks`. So MIT iconify icons look fine as a pack the user brings, and bundling third-party icon artwork into mermaid itself runs against that direction regardless of licence.

**And that applies to me too.** My pack inlines **30** BPMN glyphs. They are drawn from the OMG notation rather than lifted from an icon set, but the file does not *say* so, where the treeView pack explicitly does. That is a gap in my PR, not a debating point - I am fixing the provenance statement.

On **(1) syntax** and **(3) direction-changing routing** I have views but no standing to arbitrate; those are the maintainers' calls, and (3) in particular got harder this week - 12.0.0 makes ELK the default layout.

## @knsv - your own question is still open

On #7699 (2026-05-28) you asked: *"This a duplicate of the existing issue #2623. Can we merge them?"* Nobody replied, and #7699 is still open, still `Status: Approved`, still `Required Grooming`, in parallel with this one.

For what it is worth: **#2623 holds the votes and the scoping history, #7699 holds @andreas-emrich's worked syntax proposal.** Folding #7699's design content into here and keeping this as the single BPMN issue would lose nothing and would stop the discussion splitting three ways.

## What would actually unblock this

`Required Grooming` means "needs discussion before implementation started". It has been on #7699 since May. Three of us have now started anyway, which is the predictable outcome. One decision would settle it:

**Core diagram type, or external plugin?** @sidharthv96 (#6305), @ashishjain0512 (#6305, #6317) and @pbrolin47 (#7624) have each said this is an architectural decision maintainers make per diagram. If the answer is "external plugin", @OKHP3 already has one published and I will say so on my tracker and stop. If it is "core", then the syntax question needs one owner rather than three.

Either answer is genuinely useful. The current state - three implementations, one name, no decision - is the only one that is not.

cc @andreas-emrich @OKHP3 @derari @costis-t

**Comment by filipsajdak** [comment]
@knsv following up on your Level 1/2/3 scoping above - a proposal for what comes after the open `bpmn-beta` stack (tracked in #8160), so you can say yes, no or "not yet" before any of it is written.

The open PRs cover Level 1 plus typed events and event-based gateways. I would add the rest in stages, each as small stacked PRs, and start a stage only once you are happy with the one before:

- **A - rest of Level 2:** activity markers, non-interrupting boundary events, expanded / event sub-processes, transactions, default and conditional flows, data state, nested lanes, groups across pools, semantic validation, and `classDef` / `style` / `:::` styling.
- **B - choreography and conversation diagrams**, in the same `bpmn-beta` type.
- **C - BPMN XML import/export** (see the question below).

Syntax follows what upstream already does, with no new operators:

- the parser stays Chevrotain, co-located, following the usecase reference in `new-diagram.md`
- per-element and per-edge attributes use `@{ }`: `user task t1 "Retry"@{ loop: standard }`, `gw d1@--> manual` with `d1@{ default: true }`
- flowchart's shorthands work: `split --> pick & pack`, inline declarations (`start s1 "Order" --> user task t1 "Check"`), `-->|label|`, `-. label .->`, `--->`, `~~~`
- the keyword stays `bpmn-beta` until the syntax settles

Two questions:

1. **XML:** should BPMN 2.0 XML import/export (with DI, so a diagram opens in bpmn.io / Camunda Modeler) live in this repo as a package, or outside it as a separate npm package? The core would only need to expose the laid-out geometry.
2. **Keyword:** the external plugin `@okhp3/mermaid-diagram-bpmn` also registers `bpmn-beta`. Is that a concern for you, or fine while both are beta?

**Comment by abitrolly** [comment]
> So MIT iconify icons look fine as a pack the user brings, and bundling third-party icon artwork into mermaid itself runs against that direction regardless of licence.

This diagram looks nice. It is a pity we have to care about legal battles over vector circles. 

<img width="3426" height="1683" alt="Image" src="https://github.com/user-attachments/assets/f7276459-cf27-4962-84ec-6e8401d1fdf8" />

**Comment by AmauryVanEspen** [comment]
Would like to be sure about how to implement it with Obsidian plugin.
Currently, the only approach seems to be with bpmn.io, right ?

Le ven. 18 sept. 2026, 07:34, Anatoli Babenia ***@***.***> a
écrit :

**Comment by filipsajdak** [comment]
@AmauryVanEspen - for Obsidian today, yes: bpmn.io is the practical route. The one BPMN plugin in
Obsidian's community directory is [BPMN](https://github.com/joleaf/obsidian-bpmn-plugin)
(`bpmn-plugin`), built on bpmn-js. It renders a `.bpmn` file from your vault through a `bpmn` code
block, and it includes a modeler so you can edit the diagram inside Obsidian.

Text-based BPMN in a mermaid block is not available there yet, for two separate reasons:

- **`bpmn-beta` is not in any released mermaid.** The PRs are open (#8160 tracks them) and have not
  been reviewed, so the syntax is not settled either.
- **Obsidian ships its own copy of mermaid.** Even once this merges, it reaches your notes only when
  Obsidian updates the mermaid it bundles.

There is a possible middle route, though nobody has packaged it: @OKHP3's
[`@okhp3/mermaid-diagram-bpmn`](https://www.npmjs.com/package/@okhp3/mermaid-diagram-bpmn) is an
external diagram for mermaid 10+, and Obsidian's plugin API exposes its mermaid instance through
`loadMermaid()`. A small Obsidian plugin could register the external diagram on that instance. I
have not tried it, so treat it as an idea rather than a recipe.

If your diagrams need to open in other BPMN tools as well, the bpmn.io plugin has the advantage
that `.bpmn` is the interchange format those tools read.
